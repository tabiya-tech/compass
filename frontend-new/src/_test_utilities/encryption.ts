import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
