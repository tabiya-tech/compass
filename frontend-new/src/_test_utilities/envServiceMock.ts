jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
