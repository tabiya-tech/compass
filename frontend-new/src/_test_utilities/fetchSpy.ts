import { ExtendedRequestInit } from "src/utils/customFetch/customFetch";

export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}

// spy on authenticated fetch requests
export function setupAPIServiceSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });

  jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
  return jest.spyOn(require("src/utils/customFetch/customFetch"), "customFetch");
}

export function expectCorrectFetchRequest(fetchSpy: jest.SpyInstance, expectedUrl: string, expectedConfig: ExtendedRequestInit) {
  expect(fetchSpy).toHaveBeenCalledTimes(1);
  const [actualUrl, actualConfig] = fetchSpy.mock.calls[0];
  expect(actualUrl).toBe(expectedUrl);
  expect(actualConfig.method).toBe(expectedConfig.method);

  // Compare headers if they exist
  if(actualConfig.headers) {
    // Handle headers comparison - convert Headers object to plain object if needed
    const actualHeaders = actualConfig.headers instanceof Headers
      ? Object.fromEntries(actualConfig.headers.entries())
      : actualConfig.headers;
    expect(actualHeaders).toEqual(
      expect.objectContaining(expectedConfig.headers),
    );
  }

  // Compare bodies by parsing JSON if they exist
  if (actualConfig.body && expectedConfig.body) {
    let actualBody: unknown;
    let expectedBody: unknown;
    
    try {
      actualBody = JSON.parse(JSON.stringify(actualConfig.body));
      expectedBody = JSON.parse(JSON.stringify(expectedConfig.body));
    } catch (e) {
      console.error('Error parsing request bodies:', e);
      throw new Error('Failed to parse request bodies for comparison');
    }
    expect(actualBody).toEqual(expectedBody);
  }

  // Check any extra properties if provided
  type ExtendedConfigKey = keyof ExtendedRequestInit;
  const extraProperties: ExtendedConfigKey[] = [
    'expectedStatusCode',
    'serviceName',
    'serviceFunction',
    'failureMessage',
    'expectedContentType'
  ] as ExtendedConfigKey[];

  extraProperties.forEach((key) => {
    if (expectedConfig[key] !== undefined) {
      expect(actualConfig[key]).toEqual(expectedConfig[key]);
    }
  });
}

