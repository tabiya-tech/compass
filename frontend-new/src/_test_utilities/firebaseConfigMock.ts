import { AuthProviderIds } from "src/auth/auth.types";

const auth = jest.fn(() => ({
  signInWithCustomToken: jest.fn(),
  signInWithEmailAndPassword: jest.fn(),
  createUserWithEmailAndPassword: jest.fn(),
  signOut: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
}));

export function mockFirebaseConfig() {
  jest.spyOn(require("src/auth/firebaseConfig"), "auth").mockImplementation(auth);
}
