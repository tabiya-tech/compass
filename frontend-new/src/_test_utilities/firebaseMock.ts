import { AuthProviderIds } from "src/auth/auth.types";

const auth = {
  currentUser: {
    getIdToken: jest.fn(),
    getTokenResult: jest.fn(),
  },
  onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
  signInWithPopup: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
};

const firebase = {
  initializeApp: jest.fn(),
  auth: () => auth,
};

export default firebase;
