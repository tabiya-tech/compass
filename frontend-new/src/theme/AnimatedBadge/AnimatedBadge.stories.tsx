import { Meta, StoryObj } from "@storybook/react";
import { Box } from "@mui/material";
import AnimatedBadge from "src/theme/AnimatedBadge/AnimatedBadge";
import BadgeOutlinedIcon from "@mui/icons-material/BadgeOutlined";

const meta: Meta<typeof AnimatedBadge> = {
  title: "Components/AnimatedBadge",
  component: AnimatedBadge,
  tags: ["autodocs"],
  argTypes: {},
};
export default meta;

type Story = StoryObj<typeof AnimatedBadge>;

export const Shown: Story = {
  args: {
    badgeContent: 2,
    invisible: false,
    children: <BadgeOutlinedIcon />,
  },
  render: (args) => (
    <Box sx={{ m: 8 }}>
      <AnimatedBadge {...args}>{args.children}</AnimatedBadge>
    </Box>
  ),
};

export const ShownWithTwoNumber: Story = {
  args: {
    badgeContent: 10,
    invisible: false,
    children: <BadgeOutlinedIcon />,
  },
  render: (args) => (
    <Box sx={{ m: 8 }}>
      <AnimatedBadge {...args}>{args.children}</AnimatedBadge>
    </Box>
  ),
};

export const ShownWithThreeNumber: Story = {
  args: {
    badgeContent: 100,
    invisible: false,
    children: <BadgeOutlinedIcon />,
  },
  render: (args) => (
    <Box sx={{ m: 8 }}>
      <AnimatedBadge {...args}>{args.children}</AnimatedBadge>
    </Box>
  ),
};
