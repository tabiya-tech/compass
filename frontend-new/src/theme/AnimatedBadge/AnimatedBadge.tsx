import React from "react";
import { Badge, keyframes, styled } from "@mui/material";

const uniqueId = "a348462e-a393-4445-9a3b-c9603e878ce7";

export const DATA_TEST_ID = {
  ANIMATED_BADGE: `animated-badge-${uniqueId}`,
};

interface AnimatedBadgeProps {
  children: React.ReactNode;
  badgeContent: number;
  invisible: boolean;
}

const pulse = keyframes`
  0% {
    transform: scale(1);
    box-shadow: 0 0 0 0 rgba(0, 255, 145, 0.8);
  }
  50% {
    transform: scale(1.1);
    box-shadow: 0 0 0 10px rgba(0, 255, 145, 0);
  }
  100% {
    transform: scale(1);
    box-shadow: 0 0 0 0 rgba(0, 255, 145, 0);
  }
`;

const StyledBadgeContent = styled("span")(({ theme }) => ({
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
  padding: `${theme.fixedSpacing(theme.tabiyaSpacing.sm)} ${theme.fixedSpacing(theme.tabiyaSpacing.xs)}`,
  borderRadius: theme.tabiyaRounding.xl * 5, // xl wasn't quite big enough when badgeContent was 99+ and full was too big, 5 was picked arbitrarily
  fontSize: theme.typography.subtitle2.fontSize,
  minWidth: theme.fixedSpacing(theme.tabiyaSpacing.lg),
  height: theme.fixedSpacing(theme.tabiyaSpacing.lg),
  backgroundColor: theme.palette.primary.main,
  animation: `${pulse} 1.5s infinite`,
}));

const AnimatedBadge: React.FC<AnimatedBadgeProps> = ({ children, badgeContent, invisible }) => {
  const badgeDisplayContent = badgeContent > 99 ? "99+" : badgeContent;

  return (
    <Badge
      badgeContent={<StyledBadgeContent>{badgeDisplayContent}</StyledBadgeContent>}
      invisible={invisible}
      data-testid={DATA_TEST_ID.ANIMATED_BADGE}
    >
      {children}
    </Badge>
  );
};

export default AnimatedBadge;
