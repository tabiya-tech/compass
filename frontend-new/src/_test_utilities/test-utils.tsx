// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
