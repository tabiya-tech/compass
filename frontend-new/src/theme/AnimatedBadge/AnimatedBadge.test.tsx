// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import AnimatedBadge, { DATA_TEST_ID } from "src/theme/AnimatedBadge/AnimatedBadge";
import BadgeOutlinedIcon from "@mui/icons-material/BadgeOutlined";

describe("AnimatedBadge", () => {
  test("should render the badge with provided content", () => {
    // GIVEN the badge content
    const givenBadgeContent = 2;

    // WHEN the component is rendered
    render(
      <AnimatedBadge badgeContent={givenBadgeContent} invisible={false}>
        <BadgeOutlinedIcon data-testid="badge-icon" />
      </AnimatedBadge>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND expect the badge to be displayed
    const badge = screen.getByTestId(DATA_TEST_ID.ANIMATED_BADGE);
    expect(badge).toBeInTheDocument();
    // AND expect the badge icon to be displayed
    expect(screen.getByTestId("badge-icon")).toBeInTheDocument();
    // AND expect the badge content to be displayed
    expect(screen.getByText(givenBadgeContent)).toBeInTheDocument();
    // AND to match the snapshot
    expect(badge).toMatchSnapshot();
  });

  test("should render the badge for a large number", () => {
    // GIVEN the badge content
    const givenBadgeContent = 100;

    // WHEN the component is rendered
    render(
      <AnimatedBadge badgeContent={givenBadgeContent} invisible={false}>
        <BadgeOutlinedIcon data-testid="badge-icon" />
      </AnimatedBadge>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND expect the badge to be displayed
    const badge = screen.getByTestId(DATA_TEST_ID.ANIMATED_BADGE);
    expect(badge).toBeInTheDocument();
    // AND expect the badge icon to be displayed
    expect(screen.getByTestId("badge-icon")).toBeInTheDocument();
    // AND expect the badge content to be displayed
    expect(screen.getByText("99+")).toBeInTheDocument();
    // AND to match the snapshot
    expect(badge).toMatchSnapshot();
  });

  test("should render the badge as invisible", () => {
    // GIVEN the badge content
    const givenBadgeContent = 2;
    // AND the badge is invisible
    const givenInvisible = true;

    // WHEN the component is rendered
    render(
      <AnimatedBadge badgeContent={givenBadgeContent} invisible={givenInvisible}>
        <BadgeOutlinedIcon data-testid="badge-icon" />
      </AnimatedBadge>
    );

    // THEN expect badge to be displayed
    const badge = screen.getByTestId(DATA_TEST_ID.ANIMATED_BADGE);
    expect(badge).toBeInTheDocument();
    // AND the badge icon to be displayed
    expect(screen.getByTestId("badge-icon")).toBeInTheDocument();
    // AND the badge content to be hidden
    expect(screen.queryByText(givenBadgeContent)).not.toBeInTheDocument();
  });
});
