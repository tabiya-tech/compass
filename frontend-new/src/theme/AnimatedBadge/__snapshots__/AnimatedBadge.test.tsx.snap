// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedBadge should render the badge for a large number 1`] = `
<span
  class="MuiBadge-root css-1c32n2y-MuiBadge-root"
  data-testid="animated-badge-a348462e-a393-4445-9a3b-c9603e878ce7"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
    data-testid="badge-icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M14 12h4v1.5h-4zm0 3h4v1.5h-4z"
    />
    <path
      d="M20 7h-5V4c0-1.1-.9-2-2-2h-2c-1.1 0-2 .9-2 2v3H4c-1.1 0-2 .9-2 2v11c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2m-9 0V4h2v5h-2zm9 13H4V9h5c0 1.1.9 2 2 2h2c1.1 0 2-.9 2-2h5z"
    />
    <circle
      cx="9"
      cy="13.5"
      r="1.5"
    />
    <path
      d="M11.08 16.18c-.64-.28-1.34-.43-2.08-.43s-1.44.15-2.08.43c-.56.24-.92.78-.92 1.39V18h6v-.43c0-.61-.36-1.15-.92-1.39"
    />
  </svg>
  <span
    class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular css-9cqolc-MuiBadge-badge"
  >
    <span
      class="css-1tmou61"
    >
      99+
    </span>
  </span>
</span>
`;

exports[`AnimatedBadge should render the badge with provided content 1`] = `
<span
  class="MuiBadge-root css-1c32n2y-MuiBadge-root"
  data-testid="animated-badge-a348462e-a393-4445-9a3b-c9603e878ce7"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
    data-testid="badge-icon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M14 12h4v1.5h-4zm0 3h4v1.5h-4z"
    />
    <path
      d="M20 7h-5V4c0-1.1-.9-2-2-2h-2c-1.1 0-2 .9-2 2v3H4c-1.1 0-2 .9-2 2v11c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2m-9 0V4h2v5h-2zm9 13H4V9h5c0 1.1.9 2 2 2h2c1.1 0 2-.9 2-2h5z"
    />
    <circle
      cx="9"
      cy="13.5"
      r="1.5"
    />
    <path
      d="M11.08 16.18c-.64-.28-1.34-.43-2.08-.43s-1.44.15-2.08.43c-.56.24-.92.78-.92 1.39V18h6v-.43c0-.61-.36-1.15-.92-1.39"
    />
  </svg>
  <span
    class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular css-9cqolc-MuiBadge-badge"
  >
    <span
      class="css-1tmou61"
    >
      2
    </span>
  </span>
</span>
`;
